File: Python/quick_sort.py
arrOfNums = [45, 62, 18, 79, 33, 50, 7, 91, 25, 84, 10, 68, 88, 96, 12, 67, 14, 89, 3, 53]


def partition(nums, lo, hi):
    pivot = nums[hi]
    i = lo - 1

    for j in range(lo, hi):
        if nums[j] <= pivot:
            i += 1
            nums[i], nums[j] = nums[j], nums[i]
    i += 1
    nums[i], nums[hi] = nums[hi], nums[i]
    return i


def quick_sort(arr, low, high):
    if low < high:
        pIndex = partition(arr, low, high)
        quick_sort(arr, low, pIndex - 1)
        quick_sort(arr, pIndex + 1, high)
    return arr


print(quick_sort(arrOfNums, 0, len(arrOfNums) - 1))

File: JS/quick_sort.ts
function qs(arr: number[], low: number, high: number) {
  if (low < high) {
    const pivotIndex = partition(arr, low, high);
    qs(arr, low, pivotIndex);
    qs(arr, pivotIndex + 1, high);
  }
  return arr;
}

function partition(nums: number[], low: number, high: number) {
  const pivot = nums[low];
  let i = low;
  let j = high;

  while (i < j) {
    while (nums[i] <= pivot) {
      i++;
    }

    while (nums[j] > pivot) {
      j--;
    }
    if (i < j) {
      const temp = nums[i];
      nums[i] = nums[j];
      nums[j] = temp;
    }
  }
  const temp = nums[j];
  nums[j] = nums[low];
  nums[low] = temp;
  return j;
}

const arrNum = [5, 1, 1, 2, 0, 0];

console.log(qs(arrNum, 0, arrNum.length - 1));
